Feature break even ratio

lomamech opened the pull request:
- Implement Integration and complete API for Break Even Reference
- Implement E2E test and complete and improve Client Side for Break Even
  Reference
- Refactor Fee Center Management, Reformulation of the Cost Center module,
  modification of the cost center and profit center assignment to a cost
  center, addition of the fixed and variable cost distinction and turnover
  of other revenues
- Modification of the expense center management module, the implementation of the break-even report by cost center, the Fee center test update, the profitability threshold and the cost breakpoint Fee Center 
closes #3549

lomamech committed:
test(Feature Break Even)
- Implement Integration and complete API for Break Even Reference
- Implement E2E test and complete and improve Client Side for Break Even
  Reference

lomamech committed:
refactor(Fee Center)
- Refactor Fee Center Management, Reformulation of the Cost Center module,
  modification of the cost center and profit center assignment to a cost
  center, addition of the fixed and variable cost distinction and turnover
  of other revenues

lomamech committed:
refactor(Fee Center)
- Modify the module as well as the cost center report, restore the display
  by combining the fixed costs of the variable costs for the cost analysis
  as well as the turnover and the other revenues for the profit analysis

lomamech committed:
test(Fee Center)
- Refactor Fee Center End to End Rest
- Update Schema and migration file

lomamech committed:
Resolve and Sanitaze Code

lomamech committed:
Resolve Conflict

lomamech committed:
Resolve conflit data.sql

lomamech committed:
Complete and Resolve conflict bhima.sql

lomamech committed:
Resolve conflict with branch Master in Data.sql

lomamech committed:
feature(BreakEven)
- Implement BreakEven By Case

lomamech committed:
feature(Break Even Ratioi)
- Clear and Sanitaze code

lomamech committed:
improvement(Break Even Report)
- Complete, clear and Sanitaze code
Fixes #3549

lomamech committed:
improvement(Break Even Report)
sanitaze and clear code

jniles commented on package.json:
```diff
@@ -58,6 +58,7 @@
     "angular-ui-bootstrap": "^2.5.6",
     "angular-ui-grid": "^4.6.6",
     "angular-ui-router": "angular-ui/angular-ui-router-bower#1.0.0-rc.1",
+    "apply": "0.0.0",
```
Where did this come from?

jniles commented on server/models/migrations/next/migrate.sql:
```diff
@@ -32,12 +46,116 @@ CREATE TABLE `bed`(
+(4, 'FORM.LABELS.BREAK_EVEN', 1);
+
+-- @lomamech 2019-03-11
+ALTER TABLE `voucher` ADD COLUMN `posted` TINYINT(1) UNSIGNED NOT NULL DEFAULT 0;
```
👍

jniles commented on client/src/modules/fee_center/modals/fee_center.modal.js:
```diff
@@ -90,17 +123,39 @@ function FeeCenterModalController($state, FeeCenter, ModalService, Notify, AppCa
-        is_cost : isCostCenter,
-      };
+      if (isVariable) {
+        vm.variableCostCenterReference = {
```
This is all the same!  The only thing that changes is the object key.  Let's rewrite this for simplicity:
```js
function onSelectAccountReference(accountReference, isCostCenter, isVariable, isTurnOver) { 
  const config = {
    account_reference_id : accountReference.id, 
    is_cost : isCostCenter, 
    is_variable : isVariable,
    is_turnover : isTurnOver,
  };

  if (isCostCenter && isVariable) {
    vm.variableCostCenterReference  = config;
  } else if (isCostCenter && !isVariable) {
    vm.fixCostCenterReference = config;
  } else if () {
  // etc..
  }

}
```

jniles commented on client/src/i18n/en/form.json:
```diff
@@ -131,15 +131,17 @@
             "PATIENT_VISIT_PENDING":"The patient has a pending visit",
             "BY": "by",
             "CANCEL_CASH": "Cancellation of payment",
+            "CAN_NOT_BREAKEVEN": "Can not calculate breakeven",
```
Let's call this "Can not calculate break even point".

jniles commented on client/src/i18n/en/form.json:
```diff
@@ -131,15 +131,17 @@
             "BY": "by",
             "CANCEL_CASH": "Cancellation of payment",
+            "CAN_NOT_BREAKEVEN": "Can not calculate breakeven",
+            "CHECK_DEFINE_FEE_CENTERS": "Check to define the fee centers to analyzed",
```
A better translation here would be "Check to determine the fee centers to be analyzed"

jniles commented on client/src/i18n/en/form.json:
```diff
@@ -392,6 +396,7 @@
             "FATHER_NAME": "Father's Name",
             "FAX": "Fax",
             "FEMALE": "Female",
+            "FIXED_CHARGE": "Fixed charge",
```
I think this should be "Fixed Cost" in English.

jniles commented on client/src/i18n/en/form.json:
```diff
@@ -571,6 +579,7 @@
             "RESULT": "Results",
             "RESULT_ACCOUNT_SCT": "Result Account section",
             "RETYPE": "Retype",
+            "REVENUE": "Revenu",
```
This should be "Revenue".

jniles commented on client/src/i18n/en/form.json:
```diff
@@ -664,6 +675,7 @@
             "UNIT_WEIGHT": "Weight",
             "UNIT_VOLUME": "Volume",
             "VALUE": "Value",
+            "VARIABLE_CHARGE": "Variable charge",
```
This should be "variable cost"

jniles commented on client/src/i18n/en/table.json:
```diff
@@ -23,10 +24,25 @@
+                "BREAK_EVEN_ANALYSIS": "Break-even analysis",
+                "BREAK_EVEN_GRAPH_ANALYSI": "Break-even Graph Analysis",
+                "BREAK_EVEN_POINT": "break even point",
+                "DETERMINATION_FIXED_VARIABLE_CHARGE" : "Determination of fixed and variable charges",
```
Here,  "charges" -> "costs"

jniles commented on client/src/i18n/en/table.json:
```diff
@@ -23,10 +24,25 @@
+                "DETERMINATION_INCOME_TURNOVER" : "Determination of income and turnover",
+                "DIFFERENTIAL_INCOME_STATEMENT" : "Differential income statement",
+                "ENLARGED_TURNOVER": "Enlarged turnover",
+                "MARGINS_VARIABLE_EXPENSES": "Margins on Variable Expenses",
```
Here, "Expenses" -> "Costs".

jniles commented on client/src/i18n/en/table.json:
```diff
@@ -215,6 +235,7 @@
             "UNPAID_INVOICE_2": " days. Please make sure that the debtor has not made a payment and schedule a recovery ",
             "USER": "User",
             "VALUE": "Value",
+            "VARIABLE_CHARGES": "Variable Charges",
```
This should be "Variable Costs"

jniles commented on client/src/i18n/en/table.json:
```diff
@@ -83,6 +100,7 @@
             "EXCHANGE_RATE": "Exchange Rate",
             "EXPIRATION_DATE": "Expiration Date",
             "EXPIRE_IN":"Expire In",
+            "FIXED_CHARGES": "Fixed charges",
```
This should be "Fixed Costs"

jniles commented on client/src/i18n/en/table.json:
```diff
@@ -23,10 +24,25 @@
+            "CASES": "Cases",
             "CANCEL_CASH": "Cancel Cash Payment",
             "CHARGE": "Charge",
+            "CHARGES": "Charges",
```
It's possible this should be "Cost", depending on where it is used.

jniles commented on client/src/i18n/fr/table.json:
```diff
@@ -175,7 +194,7 @@
             "REGISTERED_ON": "Enregistré le",
             "RESPONSIBLE": "Responsable",
-            "RESULT": "Resultat",
+            "RESULT": "Résultat",
```
💃

jniles commented on client/src/modules/break_even_reference/break_even_reference.html:
```diff
@@ -0,0 +1,45 @@
+        </button>
+      </div>
+
+      <div class="toolbar-item">
```
Let's use `bhFilterToggle` here so that it has the same behavior as all other grids.

jniles commented on client/src/modules/break_even_reference/break_even_reference.js:
```diff
@@ -0,0 +1,120 @@
+      .then((breakEvenReference) => {
+        breakEvenReference.forEach((item) => {
+          if (item.is_cost && item.is_variable) {
+            item.property = $translate.instant('FORM.LABELS.VARIABLE_CHARGE');
```
Can we call this key "label" or "hrLabel" or another word that means more than just "property"?

jniles commented on client/src/modules/break_even_reference/modals/break_even_reference.modal.html:
```diff
@@ -0,0 +1,139 @@
+    </div>
+
+    <!-- account_reference_id -->
+    <div>
```
It looks like you have an extra div here.  Let's remove it.

jniles commented on client/src/modules/break_even_reference/modals/break_even_reference.modal.html:
```diff
@@ -0,0 +1,139 @@
+    </div>
+
+    <!-- is_turnover -->
+    <div class="form-group" ng-if="!BreakEvenReferenceModalCtrl.breakEvenReferences.is_cost">
```
This should be hidden until `is_cost` is set.  Instead it is shown by default, even when nothing is set.  See below:
![RemoveThisByDefault](https://user-images.githubusercontent.com/896472/54201695-95e0cc00-44ce-11e9-9d04-7538b252f470.PNG)


A better user experience would be to show this only once a user has made the selection to which it applies.

jniles commented on client/src/modules/break_even_reference/modals/break_even_reference.modal.html:
```diff
@@ -0,0 +1,139 @@
+    <!-- is_turnover -->
+    <div class="form-group" ng-if="!BreakEvenReferenceModalCtrl.breakEvenReferences.is_cost">
+      <hr>
+      <div class="form-group">
```
No need to put `<div class="form-group">` for the checkbox.  You can remove this.

jniles commented on client/src/modules/break_even_reference/modals/break_even_reference.modal.js:
```diff
@@ -0,0 +1,80 @@
+  }
+
+  // load Break Even Reference
+  BreakEvenReference.read()
```
I don't think you use this.  You can remove it.

jniles commented on client/src/modules/break_even_reference/modals/break_even_reference.modal.js:
```diff
@@ -0,0 +1,80 @@
+  if (!vm.isCreating) {
+    BreakEvenReference.read(vm.stateParams.id)
+      .then(data => {
+        vm.breakEvenReferences = data;
```
Can we call this "breakEvenReference" (singular) or just "reference"?  It is confusing to make it plural since it is a single object.

jniles commented on client/src/modules/break_even_reference/templates/action.cell.html:
```diff
@@ -0,0 +1,30 @@
+
+  <ul data-row-menu="{{row.entity.label}}" class="dropdown-menu-right" bh-dropdown-menu-auto-dropup uib-dropdown-menu>	
+    <li class="dropdown-header text-ellipsis"  style="max-width:250px;">
+      <a> {{row.entity.label}} </a>
```
Since we cannot click this element, it doesn't make sense to wrap it in an `<a>` tag.  Let's remove that.

jniles replied:
Also, it looks like this removes the "text-ellipsis" effect:
![TextEllipsis](https://user-images.githubusercontent.com/896472/54202623-bad63e80-44d0-11e9-8213-e363bdde63c0.PNG)

jniles commented on client/src/modules/break_even_reference/modals/break_even_reference.modal.html:
```diff
@@ -0,0 +1,139 @@
+
+  <div class="modal-body" style="overflow-y: scroll; height:300px;">
+    <!-- label -->
+    <div class="form-group" ng-class="{ 'has-error' : BreakEvenReferenceForm.$submitted && BreakEvenReferenceForm.description.$invalid }">
```
This should be `BreakEvenReferenceForm.label.$invalid` instead of `BreakEvenReferenceForm.description.$invalid`.  Otherwise, it won't show validation correctly:
![CoolSongBro](https://user-images.githubusercontent.com/896472/54202472-659a2d00-44d0-11e9-8d69-15e2cba14215.PNG)

jniles commented on client/src/modules/break_even_reference/modals/break_even_reference.modal.js:
```diff
@@ -0,0 +1,80 @@
+  }
+
+  function closeModal() {
+    $state.transitionTo('break_even_reference');
```
Let's use `$state.go()` instead of `$state.transitionTo()`.

jniles commented on client/src/modules/fee_center/modals/fee_center.modal.js:
```diff
@@ -54,11 +55,24 @@ function FeeCenterModalController($state, FeeCenter, ModalService, Notify, AppCa
+
+          if (reference.is_variable) {
+            vm.feeCenter.reference_cost_variable_id = reference.account_reference_id;
+            vm.variableCostCenterReference = {
```
The `vm.variableCostCenterReference` and `vm.fixCostCenterReference` contain the same information.  When this happens, to save space, you should use an extra variable:
```js
const costCenterReferenceBundle = {
              account_reference_id : reference.account_reference_id,
              is_cost : reference.is_cost,
              is_variable : reference.is_variable,
              is_turnover : reference.is_turnover,
}

if (reference.is_variable) {
  vm.variableCostCenterReference  = costCenterReferenceBundle;
} else {
  vm.fixedCostCenterReference  = costCenterReferenceBundle;
}
```

jniles commented on client/src/modules/fee_center/modals/fee_center.modal.js:
```diff
@@ -54,11 +55,24 @@ function FeeCenterModalController($state, FeeCenter, ModalService, Notify, AppCa
+            };
+          } else {
+            vm.feeCenter.reference_cost_fixed_id = reference.account_reference_id;
+            vm.fixCostCenterReference = {
```
It looks like all of these things are the same..  why do we need to attach them to `vm` as different keys?

jniles commented on client/src/modules/home/home.html:
```diff
@@ -92,7 +92,7 @@
             <div style="margin-top : 12px">
             <div class="progress" style="margin-bottom:0px;">
-              <div class="ima-blue progress-bar" role="progressbar" style="width:{{HomeCtrl.year.percentage}}%;">
+              <div class="ima-blue progress-bar" role="progressbar" style="width:{{HomeCtrl.year.percentage}}%; color: white">
```
💃

jniles commented on server/controllers/finance/reports/break_even_fee_center/index.js:
```diff
@@ -0,0 +1,147 @@
+  const params = req.query;
+  const feeCentersChecked = [];
+
+  if (params.defineFeeCenters && params.feeCenters) {
```
Why not just check `params.feeCenters`?

jniles commented on server/controllers/finance/reports/break_even_fee_center/index.js:
```diff
@@ -0,0 +1,147 @@
+  const feeCentersChecked = [];
+
+  if (params.defineFeeCenters && params.feeCenters) {
+    params.feeCenters.forEach(item => {
```
Let's make this into an `Array.prototype.map`:
```js
feeCentersChecked = params.feeCenters.map(item => parseInt(item, 10));
```

jniles commented on server/controllers/finance/reports/break_even_fee_center/index.js:
```diff
@@ -0,0 +1,147 @@
+  params.end_date = new Date(params.end_date);
+
+  data.period = {
+    start_date : params.start_date,
```
Why not just `start_date: new Date(params.start_date),` and then you won't need to the parsing above?

lomamech replied:
I have set parsing in the line  40 and 41 before

jniles commented on server/controllers/finance/reports/break_even_fee_center/index.js:
```diff
@@ -0,0 +1,147 @@
+    fiscalYearStart : params.fiscalYearStart,
+  };
+
+  _.defaults(params, DEFAULT_PARAMS);
```
Nice 👍

jniles commented on server/controllers/finance/reports/break_even_fee_center/index.js:
```diff
@@ -0,0 +1,147 @@
+    SELECT fcd.is_cost, fcd.is_variable, fcd.is_turnover, BUID(fcd.row_uuid) AS row_uuid,
+    fca.label AS auxiliary, fcp.label AS principal, SUM(fcd.debit_equiv) AS debit,
+    SUM(fcd.credit_equiv) AS credit, gl.trans_date, fcd.auxiliary_fee_center_id,
+    fcd.principal_fee_center_id
```
Because you are grouping by `fcd.auxiliary_fee_center_id` then `fcd.principal_fee_center_id`, let's put those two first in the SELECT statement.

jniles commented on server/controllers/finance/reports/break_even_fee_center/index.js:
```diff
@@ -0,0 +1,147 @@
+    FROM patient_visit AS pv
+    JOIN patient_visit_service AS pvs ON pvs.patient_visit_uuid = pv.uuid
+    JOIN service_fee_center AS sfc ON sfc.service_id = pvs.service_id
+    JOIN fee_center AS f ON f.id = sfc.fee_center_id
```
Cool!  You made quick work adapting to #3623.

jniles commented on server/controllers/finance/reports/break_even_fee_center/index.js:
```diff
@@ -0,0 +1,147 @@
+  `;
+
+  const getEncounterFeeCenter = `
+    SELECT f.label AS feeCenter, sfc.fee_center_id, count(pv.uuid) AS numberCase 
```
Instead of `numberCase`, let's call this `numberOfCases`.

jniles commented on server/controllers/finance/reports/break_even_fee_center/index.js:
```diff
@@ -0,0 +1,147 @@
+          item.selected = 0;
+          feeCentersChecked.forEach(feeCenterId => {
+
+            if (item.id === (parseInt(feeCenterId, 10))) {
```
You already did `parseInt()` in [Line 35](https://github.com/IMA-WorldHealth/bhima/pull/3622/files#diff-dd72ea0b35f0c54fa6f66e792ad71aa3R35) of this controller, no need to do it again here.

jniles commented on server/controllers/finance/reports/break_even_fee_center/index.js:
```diff
@@ -0,0 +1,147 @@
+      const dataConfigured = setting.configuration(config);
+
+      if (feeCentersChecked.length) {
+        dataConfigured.principal.forEach(item => {
```
Looks like you are trying to get `dataConfigured.principalSelected` at the end of this operation. and that is the only reason you need `item.selected`.  I think this can be done more simply like this:
```js
dataConfigured.prinicpalSelected = dataConfigured.principal
  .filter(item => feeCentersChecked.includes(item.id));
```

jniles commented on server/controllers/finance/reports/break_even_fee_center/report.handlebars:
```diff
@@ -0,0 +1,313 @@
+                  </i>                  
+                </td>
+                <td class="text-right" style="width: 20%; font-size: 15px">
+                  {{debcred balanceVariableCost metadata.enterprise.currency_id}}
```
Are you sure this `metadata.enterprise.currency_id` is respected?  I think it should be `../metadata.enterprise.currency_id`.  Try changing the currency and see if it is.

lomamech replied:
It's possible, i need Vanga database to do this task well, if this problem is generalized in another report, maybe it will require the creation of an issue

jniles commented on server/models/migrations/next/migrate.sql:
```diff
@@ -1,9 +1,23 @@
 -- change the pavillion table to ward table
-RENAME TABLE pavillion TO ward;
+-- RENAME TABLE pavillion TO ward;
```
It looks like you decided to create a new table for ward.  Can we change this to `drop table pavillion` then?

lomamech replied:
I made this update, just because the table pavillon had already been renamed by ward even in the Tshikaji database and Branch Master

jniles commented on server/models/migrations/next/migrate.sql:
```diff
@@ -1,9 +1,23 @@
-INSERT INTO unit VALUES 
-  (228, 'Ward Configurations', 'TREE.WARD_CONFIGURATION', 'Ward configuration module', 227, '/modules/ward/configuration', '/ward/configuration');
+INSERT INTO unit VALUES
+  (227, 'Ward Module', 'TREE.WARD', 'Ward folder', 0, '/modules/ward/configuration', '/WARD_FOLDER'),
```
Great fixes here.

jniles commented on yarn.lock:
```diff
@@ -387,6 +387,11 @@ append-field@^1.0.0:
   resolved "https://registry.yarnpkg.com/append-field/-/append-field-1.0.0.tgz#1e3440e915f0b1203d23748e78edd7b9b5b43e56"
   integrity sha1-HjRA6RXwsSA9I3SOeO3XubW0PlY=
 
+apply@0.0.0:
```
You'l need to remove this as well by rerunning `yarn`.

jniles reviewed (commented):
@lomamech I've done a review.  This was a lot of code to wade through.

I've made some suggestions and comments.  Considering how important this analysis to our clients .. is there a way to do more of the summing in SQL?  Javascript is notoriously bad at doing sums correctly.  We've tried many times to get away from it, and I feel it is really important to make sure this report is correct.

lomamech committed:
Resolve conflict bhima.sql

lomamech committed:
improvement(BreakEven)
- Improve translation key
- Rewrite code for simplicity
- Set hidden the cashbox is turnOver
- Complete the review of code

lomamech committed:
improvement(BreakEven)
- Clear and Sanitaze code

lomamech committed:
test(Break Even)
- fix and Upgrate end to end test for BreakEven Reference

lomamech committed:
Set semicolon in fee center modal

lomamech committed:
fix(Report)
- Fix the using of metadata.enterprise.currency_id in Fee Center Reports,
  Break Even Report and Break Even Report By Fee Center,
- Fix the formul for the calcalation of BreakEven

lomamech committed:
Resolve Conflict with Branch master

lomamech committed:
Resolve Conflict

lomamech committed:
Fix And Resolve Conflict

lomamech committed:
Complete review of code

jniles commented on client/src/modules/break_even_reference/modals/break_even_reference.modal.html:
```diff
@@ -0,0 +1,137 @@
+            type="radio"
+            name="is_cost"
+            ng-value="1"
+            ng-click="BreakEvenReferenceModalCtrl.inSetRevenueType()"
```
Just FYI - in English, we either use "unset" or "reset" to indicate the opposite of "set" instead of "inset", which means to insert something.

jniles commented on server/controllers/finance/reports/fee_center/report.handlebars:
```diff
@@ -45,7 +45,7 @@
                 <td class="text-right"> {{add @index 1}} </td>
                 <td> {{ label }} </td>
-                <td class="text-right"> {{debcred balance metadata.enterprise.currency_id}} </td>
+                <td class="text-right"> {{debcred cost_balance ../metadata.enterprise.currency_id}} </td>
```
:dancer:

jniles reviewed (approved):
LGTM.  Great work @lomamech!

bors r+
